pkg/actor/actor.go

```diff
@@ -0,0 +1,518 @@
+		metricProcBatch:       batchSizeHistogram.WithLabelValues(b.name, "proc"),
+		metricMsgBatch:        batchSizeHistogram.WithLabelValues(b.name, "msg"),
+	}, router
+}
+
+// System is the runtime of Actors.
+type System struct {
+	name                 string
+	numWorker            int
+	actorBatchSize       int
+	msgBatchSizePerActor int
+
+	rd     *ready
+	router *Router
+	wg     *errgroup.Group
+	cancel context.CancelFunc
+
+	fatalHandler func(string, ID)
+
+	// Metrics
+	metricTotalWorkers    prometheus.Gauge
+	metricWorkingWorkers  prometheus.Gauge
+	metricWorkingDuration prometheus.Counter
+	metricPollDuration    prometheus.Observer
+	metricProcBatch       prometheus.Observer
+	metricMsgBatch        prometheus.Observer
+}
+
+// Start the system. Cancelling the context to stop the system.
+// Start is not threadsafe.
+func (s *System) Start(ctx context.Context) {
+	s.wg, ctx = errgroup.WithContext(ctx)
+	ctx, s.cancel = context.WithCancel(ctx)
+
+	s.metricTotalWorkers.Add(float64(s.numWorker))
+	for i := 0; i < s.numWorker; i++ {
+		s.wg.Go(func() error {
+			s.poll(ctx)
+			return nil
+		})
+	}
+}
+
+// Stop the system, cancels all actors. It should be called after Start.
+// Stop is not threadsafe.
+func (s *System) Stop() error {
+	s.metricTotalWorkers.Add(-float64(s.numWorker))
+	if s.cancel != nil {
+		s.cancel()
+	}
+	s.rd.stop()
+	return s.wg.Wait()
+}
+
+// Spawn spawns an actor in the system.
+// Spawn is threadsafe.
+func (s *System) Spawn(mb Mailbox, actor Actor) error {
+	id := mb.ID()
+	p := &proc{mb: mb, actor: actor}
+	return s.router.insert(id, p)
+}
+
+const slowReceiveThreshold = time.Second
+
+// The main poll of actor system.
+func (s *System) poll(ctx context.Context) {
+	batchP := make([]*proc, 0, s.actorBatchSize)
+	batchMsg := make([]pipeline.Message, 0, s.msgBatchSizePerActor)
+	s.metricProcBatch.Observe(float64(len(batchP)))
+	rd := s.rd
+	rd.Lock()
+
+	startTime := time.Now()
+	s.metricWorkingWorkers.Add(1)
+	for {
+		batchP = batchP[:0]
+		for {
+			if rd.stopped {
+				rd.Unlock()
+				return
+			}
+			for i := 0; i < s.actorBatchSize; i++ {
+				if rd.queue.Len() == 0 {
+					break
+				}
+				element := rd.queue.Front()
+				rd.queue.Remove(element)
+				p := element.Value.(*proc)
+				batchP = append(batchP, p)
+			}
+			if len(batchP) != 0 {
+				break
+			}
+			// Recording metrics.
+			s.metricWorkingDuration.Add(time.Since(startTime).Seconds())
+			s.metricWorkingWorkers.Add(-1)
+			rd.cond.Wait()
+			startTime = time.Now()
+			s.metricWorkingWorkers.Add(1)
+		}
+		rd.Unlock()
+
+		for _, p := range batchP {
+			batchMsg = batchMsg[:0]
+			for i := 0; i < s.msgBatchSizePerActor; i++ {
+				msg, ok := p.mb.tryReceive()
+				if !ok {
+					break
+				}
+				batchMsg = append(batchMsg, msg)
+			}
+			if len(batchMsg) == 0 {
+				continue
+			}
+			s.metricMsgBatch.Observe(float64(len(batchMsg)))
+
+			pollStartTime := time.Now()
+			close := !p.actor.Poll(ctx, batchMsg)
```

Comment:
close is the builtin function name, use another variable name

Reply:
this pieces of code is quite long and makes others feel complicate, extract some method to make it easier to read.